### PR TITLE
uReplicator Worker with high level consumer

### PR DESCRIPTION
**Pull request by yangy0000**
- convert uReplicator Worker to java code
- use high-level consumer to support security and header
- simplify the layering inside uReplicator worker
- use multiple producers to resolve produce throughput bottleneck
 
sample startup command:
**java -Dlog4j.configuration=file:config/test-log4j.properties -XX:MaxGCPauseMillis=100 -XX:InitiatingHeapOccupancyPercent=45 -verbose:gc -Xmx1g -Xms1g -XX:+UseG1GC -XX:+PrintGCDetails -XX:+PrintGCTimeStamps -XX:+PrintGCDateStamps -XX:+PrintTenuringDistribution -Xloggc:gc-ureplicator-worker.log -server -Dcom.sun.management.jmxremote -Dcom.sun.management.jmxremote.port=9089 -Dcom.sun.management.jmxremote.rmi.port=9089 -Dcom.sun.management.jmxremote.local.only=false -Dcom.sun.management.jmxremote.authenticate=false -Dcom.sun.management.jmxremote.ssl=false -Djava.rmi.server.hostname=localhost -cp uReplicator-Worker-3.0/target/uReplicator-Worker-3.0-2.0.0-SNAPSHOT-jar-with-dependencies.jar com.uber.stream.ureplicator.worker.WorkerStarter -federated_enabled false -consumer_config config/consumer.properties -producer_config config/producer.properties -helix_config config/helix.properties**

### CODE REVIEW

**Comment by zhanglu9** [comment]
How to config consumer to use SSL?
I have a consumer-ssl.properties file like the following and when start the worker
`bootstrap.servers=pk1:443,pk2:443
security.protocol=SSL
ssl.truststore.location=/srv/kafka/config/server.truststore-chain.jks
ssl.truststore.password=password1234
ssl.keystore.location=/srv/kafka/config/server.keystore.jks
ssl.keystore.password=password1234
ssl.key.password=password1234
ssl.endpoint.identification.algorithm=
zookeeper.connect=pk1:2181,pk2:2181,pk3:2181/kafka-bf1
group.id=uReplicator-mirrormaker-bf1
consumer.id=kloakmms01-sjc1
socket.receive.buffer.bytes=1048576
fetch.message.max.bytes=8388608
queued.max.message.chunks=5
auto.offset.reset=largest`
the worker startup script like following 
`
exec "$JAVACMD" $JAVA_OPTS -Dapp_name=uReplicator-Worker \
  -classpath "$CLASSPATH" \
  -Dapp.name="start-worker-auto" \
  -Dapp.pid="$$" \
  -Dapp.repo="$REPO" \
  -Dapp.home="$BASEDIR" \
  -Dbasedir="$BASEDIR" \
  com.uber.stream.kafka.mirrormaker.starter.MirrorMakerStarter \
  startMirrorMakerWorker kafka.mirrormaker.MirrorMakerWorker \
                  --consumer.config config/consumer-ssl.properties \
                  --producer.config config/producer-ssl.properties \
                  --helix.config config/helix.properties "$@"
`
when start the worker, it first complains about all those "ssl.xxx" are not valid properties, and then 
`[2019-06-24 15:06:31,022] WARN [uReplicator-mirrormaker-bf1_kloakmms01-sjc1-leader-finder-thread]: Failed to find leader for Set(topic1-0, topic2-0) (kafka.mirrormaker.CompactConsumerFetcherManager$LeaderFinderThread:72)
kafka.common.BrokerEndPointNotAvailableException: End point with security protocol PLAINTEXT not found for broker 1
	at kafka.client.ClientUtils$$anonfun$getPlaintextBrokerEndPoints$1$$anonfun$apply$5.apply(ClientUtils.scala:149)
	at kafka.client.ClientUtils$$anonfun$getPlaintextBrokerEndPoints$1$$anonfun$apply$5.apply(ClientUtils.scala:149)
	at scala.Option.getOrElse(Option.scala:121)
	at kafka.client.ClientUtils$$anonfun$getPlaintextBrokerEndPoints$1.apply(ClientUtils.scala:149)
	at kafka.client.ClientUtils$$anonfun$getPlaintextBrokerEndPoints$1.apply(ClientUtils.scala:145)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
	at scala.collection.mutable.ResizableArray$class.foreach(ResizableArray.scala:59)
	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:48)
	at scala.collection.TraversableLike$class.map(TraversableLike.scala:234)
	at scala.collection.AbstractTraversable.map(Traversable.scala:104)
	at kafka.client.ClientUtils$.getPlaintextBrokerEndPoints(ClientUtils.scala:145)
	at kafka.mirrormaker.CompactConsumerFetcherManager$LeaderFinderThread.doWork(CompactConsumerFetcherManager.scala:345)
	at kafka.utils.ShutdownableThread.run(ShutdownableThread.scala:82)
[2019-06-24 15:06:31,023] INFO [CompactConsumerFetcherManager-1561388781888] Added fetcher for partitions ArrayBuffer() (kafka.mirrormaker.CompactConsumerFetcherManager:66)
[2019-06-24 15:06:31,224] INFO [uReplicator-mirrormaker-bf1_kloakmms01-sjc1-leader-finder-thread]: Partitions without leader Set(topic1-0, topic2-0) (kafka.mirrormaker.CompactConsumerFetcherManager$LeaderFinderThread:66)
[2019-06-24 15:06:31,230] WARN [uReplicator-mirrormaker-bf1_kloakmms01-sjc1-leader-finder-thread]: Failed to find leader for Set(process_open_sockets-0, listening_ports-0) (kafka.mirrormaker.CompactConsumerFetcherManager$LeaderFinderThread:72)
kafka.common.BrokerEndPointNotAvailableException: End point with security protocol PLAINTEXT not found for broker 1
	at kafka.client.ClientUtils$$anonfun$getPlaintextBrokerEndPoints$1$$anonfun$apply$5.apply(ClientUtils.scala:149)
	at kafka.client.ClientUtils$$anonfun$getPlaintextBrokerEndPoints$1$$anonfun$apply$5.apply(ClientUtils.scala:149)
	at scala.Option.getOrElse(Option.scala:121)
	at kafka.client.ClientUtils$$anonfun$getPlaintextBrokerEndPoints$1.apply(ClientUtils.scala:149)
	at kafka.client.ClientUtils$$anonfun$getPlaintextBrokerEndPoints$1.apply(ClientUtils.scala:145)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
	at scala.collection.mutable.ResizableArray$class.foreach(ResizableArray.scala:59)
	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:48)
	at scala.collection.TraversableLike$class.map(TraversableLike.scala:234)
	at scala.collection.AbstractTraversable.map(Traversable.scala:104)
	at kafka.client.ClientUtils$.getPlaintextBrokerEndPoints(ClientUtils.scala:145)
	at kafka.mirrormaker.CompactConsumerFetcherManager$LeaderFinderThread.doWork(CompactConsumerFetcherManager.scala:345)
	at kafka.utils.ShutdownableThread.run(ShutdownableThread.scala:82)
`
I have verified that all keystores are good, and able to produce/consume topics using the same settings with kafka client.

Thanks,

**Comment by zhanglu9** [comment]
Any ETA on the release date for this PR?

**Comment by yangy0000** [comment]
> Any ETA on the release date for this PR?

we are behind original schedule right now.  hopefully, we can release this PR in July.

**Comment by masterjeb** [comment]
Are you still looking at a July release, or is this getting pushed back to August? Thanks!

**Comment by yangy0000** [comment]
> Are you still looking at a July release, or is this getting pushed back to August? Thanks!

we are going to release this PR soon, but there is some other improvement will coming later.

File: uReplicator-Worker-3.0/pom.xml-org [hunk 1]
```diff
@@ -0,0 +1,100 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (xhl1988) [inline uReplicator-Worker-3.0/pom.xml-org]: Remove this file.

**Pull request merged**